### PR TITLE
Catalog API changes

### PR DESCRIPTION
**Pull request by pierre**

### CODE REVIEW

File: src/main/java/org/killbill/billing/catalog/api/VersionedCatalog.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.billing.catalog.api;
+
+import java.util.List;
+
+// Static catalog API methods must operate on the latest ("current") version
+public interface VersionedCatalog<CatalogType> extends StaticCatalog, Catalog {
```
Comment (sbrossie) [inline src/main/java/org/killbill/billing/catalog/api/VersionedCatalog.java]: Doesn't this conflict with our [implementation class](https://github.com/killbill/killbill/blob/killbill-0.19.16/catalog/src/main/java/org/killbill/billing/catalog/VersionedCatalog.java#L74) ?

File: src/main/java/org/killbill/billing/catalog/api/VersionedCatalog.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    public List<CatalogType> getVersions();
+
+    public void add(CatalogType catalog);
```
Comment (sbrossie) [inline src/main/java/org/killbill/billing/catalog/api/VersionedCatalog.java]: Should we export `add(CatalogType catalog);` ? In this case should we export [addAll](https://github.com/killbill/killbill/blob/killbill-0.19.16/catalog/src/main/java/org/killbill/billing/catalog/VersionedCatalog.java#L198)

File: src/main/java/org/killbill/billing/catalog/api/CatalogUserApi.java [hunk 3]
```diff
@@ -33,11 +33,12 @@
...
      *
-     * @param catalogName the name of the catalog
-     * @param context     the user context that specifies the tenant information
+     * @param catalogName        the name of the catalog
+     * @param catalogDateVersion optional date filter to retrieve a specific version
+     * @param context            the user context that specifies the tenant information
      * @return the {@code Catalog}
      */
-    Catalog getCatalog(String catalogName, TenantContext context) throws CatalogApiException;
+    VersionedCatalog<? extends StaticCatalog> getCatalog(String catalogName, DateTime catalogDateVersion, TenantContext context) throws CatalogApiException;
```
Comment (sbrossie) [inline src/main/java/org/killbill/billing/catalog/api/CatalogUserApi.java]: What is the use case for this `catalogDateVersion ` filter - should we use it to return one version or set of version after such date?

File: src/main/java/org/killbill/billing/catalog/api/VersionedCatalog.java (hunk 1 above)
Comment (sbrossie) [reply src/main/java/org/killbill/billing/catalog/api/VersionedCatalog.java]: I see you renamed `VersionedCatalog` -> `DefaultVersionedCatalog` inside KB.

File: src/main/java/org/killbill/billing/catalog/api/CatalogUserApi.java (hunk 3 above)
Comment (pierre) [reply src/main/java/org/killbill/billing/catalog/api/CatalogUserApi.java]: It performs an exact match (mainly used by Kaui).

File: src/main/java/org/killbill/billing/catalog/api/VersionedCatalog.java (hunk 2 above)
Comment (pierre) [reply src/main/java/org/killbill/billing/catalog/api/VersionedCatalog.java]: Indeed, we don't need it anymore. Fixed in c2ae4a7ea1e6d323f5974c471765676c6b243ddf.

**Pull request merged**